include/kodlab_mjbots_sdk/pid_controller.h

```diff
@@ -0,0 +1,533 @@
+    : 
+      PIDController<Scalar, N>(p_gain,d_gain,i_gain,time_step,deadband,windup_limit,saturation){}
+
+  
+  VectorNS Update(Scalar state,Scalar derivative_state) {
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
+    // converts scalar to vector form to calculate output
+    return Update(state*VectorNS::Ones(),derivative_state*VectorNS::Ones());
+  }
+
+  VectorNS Update(const VectorNS & state,const VectorNS & derivative_state) {
+
+    return Update(state,derivative_state,time_step_);
+  }  
+
+  /**
+   * @brief Function to update setpoints and PID output.
+   * @param state Current state for the controller.
+   * @param derivative_state Current derivative state of the controller.
+   * @param time_step   Time difference between 2 updates
+   * @return Returns the Pid output
+   */
+  VectorNS Update (const VectorNS & state,const VectorNS & derivative_state,double time_step) override  {
+    // Calculate the errors
+    // Proportional error
+    error_ = setpoint_ - state;
+    
+    v_cmd_ = kp_.array() *error_.array() + derivative_setpoint_.array();
+    
+    v_error_ = v_cmd_-derivative_state;
+    
+    // Calling UpdateWithError.
+    return UpdateWithError(error_, v_error_,time_step);
+  }
+
+  /**
+   * @brief Function to update the PID output given the respective errors
+   * @param error_p Proportional Error
+   * @param error_d Derivative Error
+   * @return Returns the Pid output
+   */
+  
+  VectorNS UpdateWithError(const VectorNS & error_p, const VectorNS &  error_v, double time_step) override  {
+    // Computing the integral error
+    i_error_ += error_v*time_step;
+    
+    //adding a limit on integral error to prevent windup
+    i_error_ = i_error_.cwiseMin(windup_limit_).cwiseMax(-windup_limit_);
+
+    // Finding the control output from the errors
+    output_ = kd_.array() * error_v.array() + ki_.array() * i_error_.array();
+     
+    //Checking if in Deadband range
+    for(int i=0;i<error_p.rows();i++){
+      if ( std::abs(error_p.coeff(i)) < deadband_.coeff(i)) {
+          output_(i) =0; 
+          i_error_(i)=0;
+      } 
+    }
+    
+    return output_;
+  }
+
+protected:
+  using PIDController<Scalar,N>::kp_;
+  using PIDController<Scalar,N>::ki_;
+  using PIDController<Scalar,N>::kd_;
+  using PIDController<Scalar,N>::error_;
+  using PIDController<Scalar,N>::i_error_;
+  using PIDController<Scalar,N>::time_step_;
+  using PIDController<Scalar,N>::deadband_;
+  using PIDController<Scalar,N>::output_;
+  using PIDController<Scalar,N>::windup_limit_;
+  using PIDController<Scalar,N>::setpoint_;
+  using PIDController<Scalar,N>::derivative_setpoint_;
+
+  /**
+   * @brief Stores the previous error for the d term
+   */
+  VectorNS v_error_ ;
+
+  /**
+   * @brief Stores the previous error for the d term
+   */
+  VectorNS v_cmd_ ;
+};
+
```

Comment:
Should add a VectorNS max_velocity_;  and accompanying setters and getters and than clamp v_cmd_ in the v_error_ calculation or the v_cmd_ we save.